### ISSUE TITLE
Loss Returning NaN Values at 60K Iterations with float16 on V100 GPUs

### ISSUE THREAD

**Issue by ryanbty** [karpathy/nanoGPT#383]
Hi everyone, 

I am training a ```GPT-2 (124M parameters)``` on ```OpenWebText``` from scratch on four ```V100 GPUs```  using ```float16``` precision and the ```AdamW ``` optimizer. The V100 GPUs do not support ```bfloat16```. All goes well until around ```60,000 iterations```, where the loss starts returning ```NaN values```  despite gradient scaling working well with float16.

Would appreciate any recommendations on learning rate or other settings to prevent this issue. Thanks!

**Comment by shubhamkakde111** [comment]
hello sir, 
     I'm Getting the same issue when im trying to train nanGPT using V100 GPUs but at very early stage  like 120 iterations im getting NaN
. I'm training it for 5000 iterations and in a distributed way using 4 nodes where 1 process per node using MPI backend.
following are the logs im getting : 
           train a miniature character-level shakespeare model
 good for debugging and playing on macbooks and such

out_dir = 'out-shakespeare-char'
eval_interval = 250 # keep frequent because we'll overfit
eval_iters = 200
log_interval = 10 # don't print too too often

 we expect to overfit on this small dataset, so only save when val improves
always_save_checkpoint = False

wandb_log = False # override via command line if you like
wandb_project = 'shakespeare-char'
wandb_run_name = 'mini-gpt'

dataset = 'shakespeare_char'
gradient_accumulation_steps = 4
batch_size = 64
block_size = 256 # context of up to 256 previous characters

   baby GPT model :)
  n_layer = 6
  n_head = 6
  n_embd = 384
  dropout = 0.2
  
  learning_rate = 1e-3 # with baby networks can afford to go a bit higher
  max_iters = 5000
  lr_decay_iters = 5000 # make equal to max_iters usually
  min_lr = 1e-4 # learning_rate / 10 usually
  beta2 = 0.99 # make a bit bigger because number of tokens per iter is small
  
  warmup_iters = 100 # not super necessary potentially
  
   on macbook also add
  device = 'cpu'  # run on cpu only
   compile = False # do not torch compile the model
  
  mpi
  [rank 0] before compile: building model on device cuda:0
  tokens per iteration will be: 262,144
  found vocab_size = 65 (inside data/shakespeare_char/meta.pkl)
  Initializing a new model from scratch
  number of parameters: 10.65M
  num decayed parameter tensors: 26, with 10,740,096 parameters
  num non-decayed parameter tensors: 13, with 4,992 parameters
  using fused AdamW: True
  compiling the model... (takes a ~minute)
  step 0: train loss 4.2874, val loss 4.2823
  iter 0: loss 4.2667, time 26405.90ms, mfu -100.00%
  iter 10: loss 3.2609, time 629.03ms, mfu 2.37%
  iter 20: loss 2.8381, time 634.22ms, mfu 2.37%
  iter 30: loss 2.6307, time 639.31ms, mfu 2.36%
  iter 40: loss 2.5871, time 643.02ms, mfu 2.36%
  iter 50: loss 2.5590, time 640.18ms, mfu 2.36%
  iter 60: loss 2.5022, time 641.48ms, mfu 2.35%
  iter 70: loss 2.7744, time 650.01ms, mfu 2.35%
  iter 80: loss 3.1376, time 646.98ms, mfu 2.34%
  iter 90: loss 3.2637, time 643.61ms, mfu 2.34%
  iter 100: loss 3.2875, time 649.18ms, mfu 2.34%
  iter 110: loss 3.3261, time 642.61ms, mfu 2.33%
  iter 120: loss nan, time 648.16ms, mfu 2.33%
  iter 130: loss nan, time 645.02ms, mfu 2.33%
  iter 140: loss nan, time 650.07ms, mfu 2.32%
  iter 150: loss nan, time 651.59ms, mfu 2.32%
  iter 160: loss nan, time 646.81ms, mfu 2.32%
  iter 170: loss nan, time 644.43ms, mfu 2.32%
  iter 180: loss nan, time 649.80ms, mfu 2.32%
  iter 190: loss nan, time 654.63ms, mfu 2.31%
  iter 200: loss nan, time 650.35ms, mfu 2.31%
 any help would be highly appreciated.
